pennylane/_grad.py

```diff
@@ -42,9 +45,40 @@ def _capture_diff(func, argnum=None, diff_prim=None, method=None, h=None):
 
     @wraps(func)
     def new_func(*args, **kwargs):
-        jaxpr = jax.make_jaxpr(partial(func, **kwargs))(*args)
-        prim_kwargs = {"argnum": argnum, "jaxpr": jaxpr.jaxpr, "n_consts": len(jaxpr.consts)}
-        return diff_prim.bind(*jaxpr.consts, *args, **prim_kwargs, method=method, h=h)
+        flat_args, in_trees = zip(*(tree_flatten(arg) for arg in args))
+        full_in_tree = treedef_tuple(in_trees)
+
+        # Create a new input tree that only takes inputs marked by argnum into account
+        trainable_in_trees = (in_tree for i, in_tree in enumerate(in_trees) if i in argnum)
+        trainable_in_tree = treedef_tuple(trainable_in_trees)
+
+        # Create argnum for the flat list of input arrays. For each flattened argument,
+        # add a list of flat argnums if the argument is trainable and an empty list otherwise.
+        start = 0
+        flat_argnum = sum(
+            (
+                (
+                    list(range(start, (start := start + len(flat_arg))))
+                    if i in argnum
+                    else list(range((start := start + len(flat_arg)), start))
+                )
+                for i, flat_arg in enumerate(flat_args)
+            ),
+            start=[],
+        )
```

Comment:
~What's the point of setting `start=[]`?~ Did some local coding and I see why it's used. Very cool, and concise solution, though I found it a bit unreadable. I'd say `black` is partially to blame. Maybe it'll be a bit more readable if you split it up into one step to create the generator, and another to compute its sum?
```suggestion
        start = 0
        flat_argnum_gen = (
            (
                list(range(start, (start := start + len(flat_arg))))
                if i in argnum
                else list(range((start := start + len(flat_arg)), start))
            )
            for i, flat_arg in enumerate(flat_args)
        )
        flat_argnum = sum(flat_argnum_gen, start=[])
```

Reply:
Good idea! :) Yeah I must admit it is not the most readable expression :thinking: